packages/channels/botpress-channel-web/src/db.js

```diff
@@ -75,28 +93,67 @@ module.exports = knex => {
       throw new Error(`Conversation "${conversationId}" not found`)
     }
 
+    knownConvosCache.set(conversationId, userId)
+  }
+
+  async function getUserInfo(userId) {
+    userId = sanitizeUserId(userId)
+
+    let res = userInfoCache.get(userId)
+    if (res) {
+      return res
+    }
+
+    const user = await knex('users')
+      .where({ platform: 'webchat', userId })
+      .then()
+      .get(0)
+
+    res = {
+      full_name: user && `${user.first_name} ${user.last_name}`,
+      avatar_url: (user && user.picture_url) || null
+    }
+
+    userInfoCache.set(userId, res)
+
+    return res
+  }
+
+  const bumpLastHeardOn = memoize(conversationId =>
+    throttle(
+      () =>
+        knex('web_conversations')
+          .where('id', conversationId)
+          .update({ last_heard_on: h.date.now() })
+          .then(),
+      50
+    )
+  )
+
+  async function appendUserMessage(userId, conversationId, { type, text, raw, data }) {
+    userId = sanitizeUserId(userId)
+
+    await checkConversation(conversationId, userId)
+
     const message = {
       id: uuid.v4(),
       conversationId,
       userId,
-      full_name: fullName,
-      avatar_url,
+      ...(await getUserInfo(userId)),
       message_type: type,
       message_text: text,
-      message_raw: helpers(knex).json.set(raw),
-      message_data: helpers(knex).json.set(data),
-      sent_on: helpers(knex).date.now()
+      message_raw: h.json.set(raw),
+      message_data: h.json.set(data),
+      sent_on: h.date.now()
     }
 
+    // PERF: message insertion is a serious bottlenecks
```

Comment:
I am thinking we should have an in-memory batching queue (which can be switched to Redis for multi-nodes and fault-tolerant deployments) which would flush the writes periodically. Maybe let's add a comment here to keep that in mind.